covid19_sfbayarea/data/marin_scraper.py

```diff
@@ -0,0 +1,303 @@
+			raise ValueError(f'Cannot handle media type "{media}"')
+		csv_string = unquote_plus(data)
+
+	# Then leave the iframe
+	driver.switch_to.default_content()
+
+	return csv_string
+
+def get_metadata(url: str, chart_ids: str) -> Tuple:
+	notes = []
+	driver = webdriver.Chrome('/Users/angelakwon/Downloads/chromedriver') # change this to point to Github one
+	driver.implicitly_wait(30)
+	driver.get(url)
+	soup = BeautifulSoup(driver.page_source, 'html5lib')
+	metadata = []
+
+	to_be_matched = ['Total Cases, Recovered, Hospitalizations and Deaths by Date Reported', 'Daily Count of Positive Results and Total Tests for Marin County Residents by Test Date ', 'Cases, Hospitalizations, and Deaths by Age, Gender and Race/Ethnicity ']
+	chart_metadata = []
+
+	for text in to_be_matched:
+		target = soup.find('h4',text=text)
+		if not target:
+			raise ValueError('Cannot handle this header.')
+		for sib in target.find_next_siblings()[:1]: # I only want the first paragraph tag
+			# Is it more efficient to use something like (soup object).select('h1 + p') to grab the first paragraph that follows?
+			metadata += [sib.text]
+
+	# Metadata for each chart visualizing the data of the csv file I'll pull. There's probably a better way to organize this.
+	for chart_id in chart_ids.values():
+		frame = driver.find_element_by_css_selector(f'iframe[src*="//datawrapper.dwcdn.net/{chart_id}/"]')
+		driver.switch_to.frame(frame)
+		# The metadata for the charts is located in elements with the class `dw-chart-notes' 
+		for c in driver.find_elements_by_class_name('dw-chart-notes'):
+			chart_metadata.append(c.text)
+
+		# Switch back to the parent frame to "reset" the context
+		driver.switch_to.parent_frame()
+
+	driver.quit() 
+
+	# Return the metadata. I take the set of the chart_metadata since there are repeating metadata strings.
+	return metadata, list(set(chart_metadata)) 
+
+def get_case_series(chart_id: str, url: str) -> List:
+	"""This method extracts the date, number of cumulative cases, and new cases."""
+	csv_ = extract_csvs(chart_id, url)
+	series = []
+
+	csv_strs = csv_.splitlines()
+	keys = csv_strs[0].split(',')
+	
+	if keys != ['Date', 'Total Cases', 'Total Recovered*', 'Total Hospitalized', 'Total Deaths']:
+		raise ValueError('The headers have changed')
+
+	case_history = []
+
+	for row in csv_strs[1:]:
+	 	daily = {}
+	 	# Grab the date in the first column
+	 	date_time_obj = datetime.strptime(row.split(',')[0], '%m/%d/%Y')
+	 	daily["date"] = date_time_obj.isoformat()
+	 	# Collect the case totals in order to compute the change in cases per day 
+	 	case_history.append(int(row.split(',')[1]))
+	 	# Grab the cumulative number in the fifth column
+	 	daily["cumul_cases"] = int(row.split(',')[1])
+	 	series.append(daily)
+		
+	case_history_diff = np.diff(case_history) 
+	# there will be no calculated difference for the first day, so adding it in manually
+	case_history_diff = np.insert(case_history_diff, 0, 0) 
```

Comment:
This should be the total cases from the first row rather than 0, shouldn’t it? And then each row needs to have the total cases from the first row added to it.

Reply:
whoops, yes! Thanks for catching that :)